models/pulls/pull_review.go

```diff
@@ -0,0 +1,136 @@
+// Copyright 2022 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+package pulls
+
+import (
+	"fmt"
+
+	"code.gitea.io/gitea/models/db"
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/timeutil"
+)
+
+// ViewedState stores for a file in which state it is currently viewed
+type ViewedState uint8
+
+const (
+	Unviewed   ViewedState = iota
+	HasChanged             // cannot be set from the UI/ API, only internally
+	Viewed
+)
+
+func (viewedState ViewedState) String() string {
+	switch viewedState {
+	case Unviewed:
+		return "unviewed"
+	case HasChanged:
+		return "has-changed"
+	case Viewed:
+		return "viewed"
+	default:
+		return fmt.Sprintf("unknown(value=%d)", viewedState)
+	}
+}
+
+// PRReview stores for a user - PR - commit combination which files the user has already viewed
+type PRReview struct {
+	ID           int64                  `xorm:"pk autoincr"`
+	UserID       int64                  `xorm:"NOT NULL UNIQUE(pull_commit_user)"`
+	UpdatedFiles map[string]ViewedState `xorm:"NOT NULL TEXT JSON"`                          // Stores for each of the changed files of a PR whether they have been viewed, changed since last viewed, or not viewed
+	CommitSHA    string                 `xorm:"NOT NULL UNIQUE(pull_commit_user)"`           // Which commit was the head commit for the review?
+	PullID       int64                  `xorm:"NOT NULL UNIQUE(pull_commit_user) DEFAULT 0"` // Which PR was the review on?
+	UpdatedUnix  timeutil.TimeStamp     `xorm:"updated"`                                     // Is an accurate indicator of the order of commits as we do not expect it to be possible to make reviews on previous commits
+}
+
+func init() {
+	db.RegisterModel(new(PRReview))
+}
+
+// GetReview returns the PRReview with all given values prefilled, whether or not it exists in the database.
+// If the review didn't exist before in the database, it won't afterwards either.
+// The returned boolean shows whether the review exists in the database
+func GetReview(userID, pullID int64, commitSHA string) (*PRReview, bool, error) {
+	review := &PRReview{UserID: userID, CommitSHA: commitSHA, PullID: pullID}
+	has, err := db.GetEngine(db.DefaultContext).Get(review)
+	return review, has, err
+}
+
+// UpdateReview updates the given review inside the database, regardless of whether it existed before or not
+// The given map of files with their viewed state will be merged with the previous review, if present
+func UpdateReview(userID, pullID int64, commitSHA string, updatedFiles map[string]ViewedState) error {
```

Comment:
Why shouldn't we allow `commitSHA` as parameter?
We actually need it because there is the (not too unlikely) case that the user reviews an outdated version of the PR, simply because another commit was pushed since he loaded the page.
The current implementation works in this case, your solution would only work by "crippling the UI" as GitHub now does:
It doesn't allow to make comments on outdated diffs. I find that to be a pretty user-unfriendly UI.